### PR TITLE
Fix ordering of documents

### LINKED ISSUE #32004: Update the rendering: fix ordering of documents
Follow-up to https://github.com/servo/servo/pull/31505

See step two of [spec](https://html.spec.whatwg.org/multipage/#update-the-rendering), and https://github.com/servo/servo/pull/31505/files#diff-2c5948584719622eef023c56cdf90e8278971f2f2eda26e7fd0485a24c32be3cR1735

I've tried doing this in https://github.com/servo/servo/pull/31505/commits/fa48454f64f01ceff17a6959c170914d46586da4, but this resulted in various unexpected test results(which I didn't note, sorry), and so it will require some further investigation. Note that our current implementation does not order docs in any way, since it usually simply handles a message immediately when received by the script-thread.

### PR DESCRIPTION
**Pull request by Taym95**
This PR fixes two things related to the update the rendering concept:

- Updating the rendering is now done according to the document ordering requirements of the spec(see test expectation update).
- The batching of rendering opportunities is now scoped by pipeline, to avoid missing one following closure of a pipeline([details](https://github.com/servo/servo/pull/32574#issuecomment-2275379106)). 

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #32004

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

### CODE REVIEW

**Comment by gterzian** [comment]
As per the unexpected timeouts, the re-ordering must somehow filter out iframes. Note that we don't want to filter-out any, we only want to re-order them. I think the nested `filter_map` call is perhaps to blame...

File: components/script/script_thread.rs [hunk 1]
```diff
@@ -1654,7 +1654,17 @@ impl ScriptThread {
             .documents
             .borrow()
             .iter()
+            .filter(|(_, document)| document.window().is_top_level())
             .map(|(id, document)| (id, DomRoot::from_ref(&*document)))
+            .flat_map(|(id, document)| {
+                let mut documents = vec![(id.clone(), document.clone())];
+                documents.extend(document.iter_iframes().filter_map(|iframe| {
+                    iframe
+                        .pipeline_id()
```
Comment (gterzian) [inline components/script/script_thread.rs:1663]: Note that our implementation of iframe is broken, and that https://github.com/servo/servo/blob/66edef806579fd0b386f4ceba473e6a9f7d0ca34/components/script/dom/htmliframeelement.rs#L462 likely returns `None` when it shouldn't. 

Perhaps a solution is to put the documents with not pipeline ids as last. (you can get the docs from the `documents` above).

**Review by gterzian — CHANGES_REQUESTED** [review]

File: components/script/script_thread.rs (hunk 1 above)
Comment (Taym95) [reply components/script/script_thread.rs:1663]: @gterzian found out that timeout it caused by `.filter(|(_, document)| !document.window().is_top_level())` not the filter_map

Comment (Taym95) [reply components/script/script_thread.rs:1663]: this is result after removing `is_top_level`, not timeout but wrong result:

image

Comment (gterzian) [reply components/script/script_thread.rs:1663]: Ok good, sounds like reversing `iter_iframes` is the way to go then!

Comment (Taym95) [reply components/script/script_thread.rs:1663]: @gterzian did you have time to look at this?

**Comment by gterzian** [comment]
So the problem is essentially https://github.com/servo/servo/issues/31973

The [test](/html/webappapis/update-rendering/child-document-raf-order.html) creates two iframes, but we actually create four documents(2 for each iframe: one "about:blank" and another), and then we remove the "about:blank" ones when the "real" ones finish navigating. I think the timeout is because the raf end-up being lost because the "about:blank" are not returned anymore as part of `iter_iframes`(while there is a time where even though they are not part of `iter_frames`, they are still inside `documents`) and then they the rafs callbacks are not called. 

So for this we actually first need to fix https://github.com/servo/servo/issues/31973

**Comment by gterzian** [comment]
Ok so actually I had another look, and the problem was rather that https://github.com/servo/servo/blob/bc1c71bd4df9424e4cceed9dc8cc38efc591095b/components/script/dom/node.rs#L3151 returns the owner doc, not the doc of the iframe...

**Comment by Taym95** [comment]
> Ok so actually I had another look, and the problem was rather that
> 
> https://github.com/servo/servo/blob/bc1c71bd4df9424e4cceed9dc8cc38efc591095b/components/script/dom/node.rs#L3151
> 
> returns the owner doc, not the doc of the iframe...

Thanks for checking this, what do you mean by owner doc? doc os iframe owner?

**Comment by jdm** [comment]
Yeah, the document that contains the iframe element.

**Comment by gterzian** [comment]
Ok I think we can merge this, perhaps @Taym95 take a look at the code I wrote, not sure if perhaps we can avoid some allocations while achieving the same result.

**Comment by Taym95** [comment]
> Ok I think we can merge this, perhaps @Taym95 take a look at the code I wrote, not sure if perhaps we can avoid some allocations while achieving the same result.

Yeah I will clean it up and let's merge it.

File: components/script/script_thread.rs [hunk 2]
```diff
@@ -1649,15 +1649,26 @@ impl ScriptThread {
...
         // but this isn't happening here.
-        let pipeline_and_docs: Vec<(PipelineId, DomRoot<Document>)> = self
+        let pipelines_to_update: Vec<PipelineId> = self
             .documents
             .borrow()
             .iter()
-            .map(|(id, document)| (id, DomRoot::from_ref(&*document)))
+            .filter(|(_, document)| document.window().is_top_level())
+            .flat_map(|(id, document)| {
+                std::iter::once(id.clone()).chain(
```
Comment (Taym95) [inline components/script/script_thread.rs:1658]: I tried to used `iter::once` here with `chain' to avoid vec allocation

Comment (Taym95) [inline components/script/script_thread.rs:1658]: @gterzian I tried to used `iter::once` here with `chain' to avoid vec allocation.

**Comment by gterzian** [comment]
I also got the below failure, and some timeouts, but only when running the test in release mode:

**Comment by gterzian** [comment]
hypothesis: we fail to run a rendering task, because the pipeline for which is was queued was closed, but after the last `note_rendering_opportunity`, which did not queue a task because one was queued for the subsequently closed pipeline.

**Review by gterzian — APPROVED** [review]
It seems that the [hypothesis](https://github.com/servo/servo/pull/32574#issuecomment-2275379106), tested by my latest commit which scopes the batching of rendering updates by pipeline, has been validated by the two latest test runs. The intermittency of the problem makes it practically impossible to validate through debugging locally. Let's merge this.

**Pull request merged**